src/core/ext/filters/client_channel/resolver/custom/custom_resolver.cc

```diff
@@ -0,0 +1,363 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Shim resolver for exposing public API for custom application resolvers.
+
+#include <grpc/support/port_platform.h>
+
+#include <limits.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <grpc/support/alloc.h>
+#include <grpc/support/string_util.h>
+
+#include "src/core/ext/filters/client_channel/lb_policy_factory.h"
+#include "src/core/ext/filters/client_channel/parse_address.h"
+#include "src/core/ext/filters/client_channel/resolver_registry.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/gpr/host_port.h"
+#include "src/core/lib/gpr/string.h"
+#include "src/core/lib/gprpp/mutex_lock.h"
+#include "src/core/lib/iomgr/closure.h"
+#include "src/core/lib/iomgr/combiner.h"
+#include "src/core/lib/iomgr/resolve_address.h"
+#include "src/core/lib/iomgr/unix_sockets_posix.h"
+#include "src/core/lib/slice/slice_internal.h"
+#include "src/core/lib/slice/slice_string_helpers.h"
+
+struct grpc_addresses {};
+struct grpc_resolver {};
+struct grpc_resolver_observer {};
+struct grpc_resolver_factory {};
+struct grpc_resolver_args {
+  const grpc_core::ResolverArgs* args;
+};
+
+namespace grpc_core {
+
+namespace {
+
+struct ResolverFactoryVTable {
+  void* (*resolve)(void* factory_user_data, grpc_resolver_args* args,
+                   grpc_resolver_observer* observer);
+  void (*destroy)(void* factory_user_data);
+};
+
+struct ResolverVTable {
+  void (*request_reresolution)(void* resolver_user_data);
+  void (*destroy)(void* resolver_user_data);
+};
+
+class CustomResolverObserver;
+
+class ResolverAddresses : public grpc_addresses {
+ public:
+  ResolverAddresses(size_t capacity) : capacity_(capacity) {
+    addresses_ = grpc_lb_addresses_create(capacity, nullptr);
+    addresses_->num_addresses = 0;
+  }
+
+  ~ResolverAddresses() { grpc_lb_addresses_destroy(addresses_); }
+
+  bool AddAddress(const char* target, bool is_balancer,
+                  const char* balancer_name) {
+    if (addresses_->num_addresses >= capacity_) {
+      return false;
+    };
+    grpc_uri* uri = grpc_uri_parse(target, false);
+    if (!uri) return false;
+    size_t index = addresses_->num_addresses++;
+    bool result = grpc_lb_addresses_set_address_from_uri(
+        addresses_, index, uri, is_balancer, balancer_name, nullptr);
+    grpc_uri_destroy(uri);
+    if (!result) {
+      addresses_->num_addresses--;
+    }
+    return result;
+  }
+
+  grpc_lb_addresses* GetAddresses() { return addresses_; }
+
+ private:
+  grpc_lb_addresses* addresses_;
+  size_t capacity_;
+};
+
+class CustomResolverFactory;
+
+class CustomResolver : public Resolver {
```

Comment:
Currently, each individual polling-based resolver needs to contain code to handle (a) automatically retrying with exponential backoff upon failure and (b) using a minimum reresolution interval to avoid hammering the name service.  However, because this functionality is actually common to all polling-based resolvers, we plan to eventually move it out of the individual resolver implementations and into the client channel code.  When we do that, we will remove this code from the individual resolver implementations, but we won't be able to do that for the plugins, since we won't control them.  So for now, I think we need to add that functionality here in the `CustomResolver` class, so that plugins can be written assuming that this functionality will be handled for them.  Later, when we move this functionality into the client channel code, we can remove it from here.

You can probably copy most of the implementation of this from src/core/ext/filters/client_channel/resolver/dns/native/dns_resolver.cc.

Reply:
Is it acceptable to update `PluginResolver` implementation to call `request_reresolution` based on the following factors?
* Direct `RequestReresolution` calls from gRPC core
* Periodic timer
* Exponential backoff on resolution failures

Can you also clarify how `Resolver` API will be changed in the future, please? I'm a bit confused how backoff it going to interact with push-based resolvers.